[Feature] Build renewal/replacement request modals

j-bunk opened the pull request:
## Notion ticket link
<!-- Please replace with your ticket's URL -->
[Renewal/Replacement Request modals](https://www.notion.so/uwblueprintexecs/Renewal-Replacement-Request-modals-71869043f61347babb2b51d10d14f450)


<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->
## Implementation description
* Created Edit modals (`EditDoctorInformationModal`, `EditPaymentDetailsModal`, `EditReasonForReplacementModal`) used in the request renewal/replacement flows

<!-- Catch all section that could be used to draw attention to anything the reviewers should keep in mind, substantial parts of your PR, anything you'd like a second opinion on, things that will be fixed in the future, or things that were left out -->
## Notes
* There is no form submission logic and validation is handled by Chakra itself. Error state will likely be addressed post-MVP.
* Left a few TODO comments in the code that should be addressed at a later time


## Checklist
- [x] My PR name is descriptive, is in imperative tense and starts with one of the following: `[Feature]`,`[Improvement]` or `[Fix]`,
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the RCD team on GitHub, or specific people who are associated with this ticket

OustanDing reviewed (changes requested):
Code looks great! Sorry there are so many comments XD this is a big PR - thanks a lot for your work. After you address my comments, feel free to just remove the buttons and the modals from the page before you commit!

j-bunk commented on components/internal/modals/EditDoctorInformationModal.tsx:
```diff
@@ -0,0 +1,143 @@
+              </Box>
+            </ModalBody>
+            <ModalFooter paddingBottom="24px" paddingTop="0px">
+              <Button colorScheme="gray" variant="solid" onClick={onClose}>
```
Code snippet from `tools/theme/components/button.ts`

``` ts
variants: {
    solid: ({ colorScheme }) => {
      const bgColor = colorScheme === 'primary' ? 'primary' : `secondary.${colorScheme}`;
      return {
        bg: bgColor,
        color: colorScheme === 'gray' ? 'black' : 'white',
        _hover: {
          bg: `overlay.${bgColor}.20`,
          _disabled: {
            bg: bgColor, // Disable color change when disabled button is hovered over
          },
        },
      };
    },
```

Based on how the button was customized, I think we pass in the color in colorScheme and it sets it the bg color as `secondary.${colorScheme}` as well as changing the colors of other things (color of the text and hover). Based on that I think that's how I'm supposed to configure it. Doing `backgroundColor="background.gray"` would change the background color without changing the text+hover colors (image below)
![image](https://user-images.githubusercontent.com/51224641/126719257-957569ab-8574-4db9-937e-04e22b889095.png)

OustanDing replied:
Okay, I see. In that case, sticking with `colorScheme=gray` is good!

j-bunk commented on components/internal/modals/EditDoctorInformationModal.tsx:
```diff
@@ -0,0 +1,143 @@
+        <ModalOverlay />
+        <form onSubmit={handleSubmit}>
+          <ModalContent paddingX="40px">
+            <ModalHeader paddingBottom="12px" paddingTop="24px">
```
Good catch! Setting it 0 made the inputs cut off a bit when they're active (first image), so I set the padding to 3px (`paddingX="3px"`) to stop that from happening (second image). Doesn't exactly match the design dimensions so lmk if you think there's smth wrong with that
![image](https://user-images.githubusercontent.com/51224641/126719364-187737be-3b5e-40e2-adab-cfa07e7f7136.png)
![image](https://user-images.githubusercontent.com/51224641/126719468-d0647561-2951-478c-a755-f4df3542bea6.png)

OustanDing replied:
Damn that's tricky. In that case, can you just make it `4px` so it's consistent with our spacing specs (multiples of 4px)?

j-bunk commented on components/internal/modals/EditDoctorInformationModal.tsx:
```diff
@@ -0,0 +1,143 @@
+            <ModalHeader paddingBottom="12px" paddingTop="24px">
+              <Text textStyle="display-medium-bold">{"Edit Doctor's Information"}</Text>
+            </ModalHeader>
+            <ModalBody paddingY="20px">
```
https://github.com/uwblueprint/richmond-centre-for-disability/pull/28#discussion_r675215214

j-bunk commented on components/internal/modals/EditDoctorInformationModal.tsx:
```diff
@@ -0,0 +1,143 @@
+                    value={addressLine2}
+                    onChange={event => setAddressLine2(event.target.value)}
+                  />
+                  <FormHelperText color="#323741">
```
Removing the `color` prop sets the color to be `#718096` so it doesn't seem to be configured yet.
I changed my code to `<FormHelperText color="text.seconday">`

j-bunk commented on components/internal/modals/EditPaymentDetailsModal.tsx:
```diff
@@ -0,0 +1,357 @@
+  const [donation, setDonation] = useState('');
+
+  //   Shipping address information state
+  const [hideShippingInfo, setHideShippingInfo] = useState(false); // Whether shipping information is visible
```
No strong opinions so I'm happy to change it to show instead, the only thing worth noting is that the value of the checkbox input would need to be changed to the opposite of the state as follows

``` tsx
<Checkbox
    paddingBottom="24px"
    isChecked={!showShippingInfo}
    onChange={event => setShowShippingInfo(event.target.checked)}
  >
    {'Same as home address'}
 </Checkbox>
 ```

OustanDing replied:
Oh, I see. In that case, we should name this variable something like `sameShippingAndHomeAddresses` as that is the original state that decides whether to show/hide the additional fields.

j-bunk replied:
Yep that makes a lot of sense!

j-bunk commented on components/internal/modals/EditPaymentDetailsModal.tsx:
```diff
@@ -0,0 +1,357 @@
+  // const [shippingFullNameInputError, setShippingFullNameInputError] = useState(''); // Error message displayed under input
+
+  //   Billing address information state
+  const [hideBillingInfo, setHideBillingInfo] = useState(false); // Whether billing information is visible
```
https://github.com/uwblueprint/richmond-centre-for-disability/pull/28#discussion_r675218445

j-bunk commented on components/internal/modals/EditPaymentDetailsModal.tsx:
```diff
@@ -0,0 +1,357 @@
+                          <Radio value="Debit">{'Debit'}</Radio>
+                          <Radio value="Cash">{'Cash'}</Radio>
+                          <Radio value="Cheque">{'Cheque'}</Radio>
+                          <Radio value="E-transfer">{'E-transfer'}</Radio>
```
Addressed! I had to make it `const [paymentMethod, setPaymentMethod] = useState<PaymentType | string>('');` instead because onChange (code snippet below) cannot accept null/undefined values and was throwing errors. Lmk if there are any issues with that


``` ts
                      <RadioGroup onChange={setPaymentMethod} value={paymentMethod}>
```

OustanDing replied:
Oh, okay. Does doing `useState<PaymentType | ''>('');` work?

j-bunk replied:
![image](https://user-images.githubusercontent.com/51224641/126722909-5fa05d02-3923-4073-80f1-d4e52c278028.png)

😬 throws an error

OustanDing replied:
Okay, let's leave it as what it is right now then!

j-bunk commented on components/internal/modals/EditPaymentDetailsModal.tsx:
```diff
@@ -0,0 +1,357 @@
+                        </Box>
+                      </FormLabel>
+                      <InputGroup>
+                        <InputLeftElement pointerEvents="none" color="#A3AEBE" fontSize="1.2em">
```
👀 Never noticed that, thanks for the tip!

j-bunk commented on components/internal/modals/EditPaymentDetailsModal.tsx:
```diff
@@ -0,0 +1,357 @@
+                        value={shippingAddressLine1}
+                        onChange={event => setShippingAddressLine1(event.target.value)}
+                      />
+                      <FormHelperText color="#323741">
```
https://github.com/uwblueprint/richmond-centre-for-disability/pull/28#discussion_r675216586

j-bunk commented on components/internal/modals/EditPaymentDetailsModal.tsx:
```diff
@@ -0,0 +1,357 @@
+                </Checkbox>
+
+                {/* Section is hidden if same as home address checkbox is checked */}
+                {!hideShippingInfo && (
```
https://github.com/uwblueprint/richmond-centre-for-disability/pull/28#discussion_r675218445

j-bunk commented:
@OustanDing Thanks for all the great feedback! I think the PR is ready for another review!
Also addressed some feedback from Charmaine which was adding Dividers and changes to capilization

OustanDing reviewed (commented):
Responded to some of your comments. Once those are addressed, it's ready to go!

OustanDing reviewed (approved):
LGTM! Great work :D :D
